### PR TITLE
Reconstruct streams via global registration, temporary impl to unblock FSDP

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117386

This is a placeholder implementation for reconstructing streams via global storage to unblock FSDP, pending proper stream support design

This PR does a few things:

1) fixes registration for devices with indices. We were only supporting "cuda", we now support "cuda:k" interfaces where k is # of gpu

2) Changes the stream objects in dynamo to take devices as device types, instead of strings, and updates the string based device APIs to gracefully take device types.

3) Introduces a reconstruct-by-global (using existing cleanup hook structures) to streams as a placeholder impl for now

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/ctx_manager.py [hunk 1]
```diff
@@ -586,6 +587,22 @@ def call_method(
...
+        assert not self.source
+        # Since we just proved that - for other such structures, like lists and dicts, reconstruction
+        # is fine and sound according to dynamo principles of treating collectives. However,
+        # streams are special in that we want to preserve the identity of the stream as the same as in the graph
+        # Normally, we would do this via codegen for the proxy mapping to an output - we cannot do this yet, as we do not
+        # yet have a plan for how we want to handle the case where the stream is used as an input or an output. Pending
+        # design, to unblock current work, we lift the stream into a global and then codegen bytecode to load it from there.
+        name = f"_stream_{self.device}_{id(self.value)}"
+        if name not in codegen.tx.output.global_scope:
+            codegen.tx.output.install_global(name, self.value)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/ctx_manager.py:602]: This uses existing cleanup hook structures.

File: torch/_dynamo/device_interface.py [hunk 2]
```diff
@@ -174,11 +174,17 @@ def get_compute_capability(device: _device_t = None):
 device_interfaces: Dict[str, Type[DeviceInterface]] = {}
 
 
-def register_interface_for_device(device: str, device_interface: Type[DeviceInterface]):
+def register_interface_for_device(
+    device: Union[str, torch.device], device_interface: Type[DeviceInterface]
+):
+    if isinstance(device, torch.device):
+        device = str(device)
```
Comment (jansel) [inline torch/_dynamo/device_interface.py:181]: I don't believe this will match the `cuda:{i}` below.

File: torch/_dynamo/device_interface.py [hunk 3]
```diff
@@ -174,11 +174,17 @@ def get_compute_capability(device: _device_t = None):
...
     device_interfaces[device] = device_interface
 
 
-def get_interface_for_device(device: str) -> Type[DeviceInterface]:
+def get_interface_for_device(device: Union[str, torch.device]) -> Type[DeviceInterface]:
+    if isinstance(device, torch.device):
+        device = str(device)
```
Comment (jansel) [inline torch/_dynamo/device_interface.py:187]: I don't believe this will match the cuda:{i} below.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/device_interface.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/device_interface.py:187]: Sorry, why not? 

If I set a breakpoint here:

image

Maybe I misunderstood your comment?

Comment (jansel) [reply torch/_dynamo/device_interface.py:187]: ah looks like you are right

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**